File: src/java/dive/src/main/java/marine/Submarine.java
package marine;

import lombok.Getter;

import java.util.List;

@Getter
public class Submarine {
    private Position position;

    public Submarine(int horizontal, int depth) {
        this.position = new Position(horizontal, depth);
    }

    public void move(List<Instruction> instructions) {
        instructions.forEach(this::move);
    }

    private void move(Instruction instruction) {
        position = switch (instruction.getText()) {
            case "down" -> position.withDepth(position.getDepth() + instruction.getX());
            case "up" -> position.withDepth(position.getDepth() - instruction.getX());
            default -> position.withHorizontal(position.getHorizontal() + instruction.getX());
        };
    }
}

File: src/ts/dive/src/Submarine.ts
import { Position } from "./Position";
import { Instruction } from "./Instruction";

export class Submarine {
    private position: Position;
    private moveTo: Map<string, Function> = new Map([
        ["down", (instruction: Instruction) => this.position.withDepth(this.position.depth + instruction.x)],
        ["up", (instruction: Instruction) => this.position.withDepth(this.position.depth - instruction.x)]
    ]);

    constructor(horizontal: number, depth: number) {
        this.position = new Position(horizontal, depth);
    }

    private newPosition(instruction: Instruction): Position {
        if(this.moveTo.has(instruction.text)) {
            return this.moveTo
                        .get(instruction.text)
                        ?.call(this, instruction);
        }
        return this.position.withHorizontal(this.position.horizontal + instruction.x);
    }

    move(instructions: Array<Instruction>): void {
        instructions.forEach(i => this.position = this.newPosition(i));
    }

    currentPosition(): Position {
        return this.position;
    }
}